watcher.go

```diff
@@ -0,0 +1,137 @@
+package watcher
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/tikv"
+	goctx "golang.org/x/net/context"
+)
+
+type Watcher struct {
+	sync.Mutex
+	client       kv.Storage
+	pollInterval time.Duration
+	callback     func(string) //should be protected by lock
+	keyName      string
+	formerValue  []byte //should be protected by lock
+	done         bool   //should be protected by lock
+}
+
+func NewWatcher(keyName string, pollInterval time.Duration, pdAddrs string) (*Watcher, error) {
+	cli, err := tikv.Driver{}.Open(fmt.Sprintf("tikv://%s", pdAddrs))
+	if err != nil {
+		return nil, err
+	}
+
+	res := Watcher{
+		client:       cli,
+		callback:     nil,
+		keyName:      keyName,
+		pollInterval: pollInterval,
+		formerValue:  make([]byte, 0),
+		done:         false,
+	}
+	go res.poll()
+	return &res, nil
+}
+
+// SetUpdateCallback sets the callback function that the watcher will call
+// when the policy in DB has been changed by other instances.
+// A classic callback is Enforcer.LoadPolicy().
+func (w *Watcher) SetUpdateCallback(callback func(string)) error {
+	w.Lock()
+	defer w.Unlock()
+	w.callback = callback
+	return nil
+}
+
+// Update calls the update callback of other instances to synchronize their policy.
+// It is usually called after changing the policy in DB, like Enforcer.SavePolicy(),
+// Enforcer.AddPolicy(), Enforcer.RemovePolicy(), etc.
+func (w *Watcher) Update() error {
+	//begin transaction
+	tx, err := w.client.Begin()
+	if err != nil {
+		return err
+	}
+	rev := 0
+	val, err := tx.Get(goctx.Background(), []byte(w.keyName))
+	if err != nil {
+		if err != kv.ErrNotExist {
+			return err
+		}
```

Comment:
``if err == kv.ErrNotExist``, do nothing?

Reply:
Exactly. If this key is not found in tikv, then it will do nothing. Under this circumstance, the variable rev will remain the initial value 0, which will fifnally be written into the key.

Reply:
Sincerely apolpgize for such a long delay. I just realized that I shut down all notification from github 3 years ago, when I first met github :)

Reply:
It doesn't matter